### PR TITLE
Virtual hosted-style support for S3 URI

### PR DESCRIPTION
**Pull request by Jokser**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Virtual hosted-style support for S3 URI

Detailed description / Documentation draft:
Enables another pattern to specify S3 URI:
https://docs.aws.amazon.com/AmazonS3/latest/dev/VirtualHosting.html#virtual-hosted-style-access

### CODE REVIEW

File: dbms/src/IO/S3Common.cpp [hunk 1]
```diff
@@ -106,46 +105,37 @@ namespace S3
...
         endpoint = uri.getScheme() + "://" + uri.getAuthority();
 
-        // Parse bucket and key from path.
-        std::smatch match;
-        std::regex_search(uri.getPath(), match, bucket_key_pattern);
-        if (!match.empty())
+        if (re2::RE2::FullMatch(uri.getAuthority(), virtual_hosted_style_pattern, &bucket))
+        {
+            if (!bucket.empty())
+                bucket = bucket.substr(0, bucket.length() - 1);
```
Comment (alexey-milovidov) [inline dbms/src/IO/S3Common.cpp:121]: pop_back?

File: dbms/src/IO/S3Common.cpp [hunk 2]
```diff
@@ -106,46 +105,37 @@ namespace S3
...
+            if (bucket.length() < 3 || bucket.length() > 63)
```
Comment (alexey-milovidov) [inline dbms/src/IO/S3Common.cpp:122]: What is the reference for these numbers (you can provide it in comment)?

File: dbms/src/IO/S3Common.cpp [hunk 3]
```diff
@@ -106,46 +105,37 @@ namespace S3
 
     URI::URI(const Poco::URI & uri_)
     {
-        static const std::regex bucket_key_pattern("([^/]+)/(.*)"); /// TODO std::regex is discouraged
+        static const RE2 virtual_hosted_style_pattern("(.+\\.)?s3[.-][a-z0-9-.]+");
```
Comment (alexey-milovidov) [inline dbms/src/IO/S3Common.cpp:108]: Maybe you can provide a reference in comments that will answer the question "what is that"?

Comment (alexey-milovidov) [reply dbms/src/IO/S3Common.cpp:108]: Don't understand:

`[.-]` means "from dot symbol to the end of byte range", is that correct?

File: dbms/src/IO/S3Common.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply dbms/src/IO/S3Common.cpp:121]: Clarification needed: why we are stripping the last character?

File: dbms/src/IO/S3Common.cpp [hunk 4]
```diff
@@ -106,46 +105,37 @@ namespace S3
...
-        std::regex_search(uri.getPath(), match, bucket_key_pattern);
-        if (!match.empty())
+        if (re2::RE2::FullMatch(uri.getAuthority(), virtual_hosted_style_pattern, &bucket))
+        {
+            if (!bucket.empty())
+                bucket = bucket.substr(0, bucket.length() - 1);
+            if (bucket.length() < 3 || bucket.length() > 63)
+                throw Exception("Invalid S3 URI bucket: " + uri.toString(), ErrorCodes::BAD_ARGUMENTS);
+
+            key = uri.getPath().substr(1);
```
Comment (alexey-milovidov) [inline dbms/src/IO/S3Common.cpp:125]: Why we are stripping the leading character?

File: dbms/src/IO/S3Common.cpp [hunk 5]
```diff
@@ -106,46 +105,37 @@ namespace S3
...
-            if (bucket.empty())
-                throw Exception ("Invalid S3 URI: no bucket: " + uri.toString(), ErrorCodes::BAD_ARGUMENTS);
+            if (bucket.length() < 3 || bucket.length() > 63)
+                throw Exception("Invalid S3 URI bucket: " + uri.toString(), ErrorCodes::BAD_ARGUMENTS);
 
-            key = match.str(2);
-            if (key.empty())
-                throw Exception ("Invalid S3 URI: no key: " + uri.toString(), ErrorCodes::BAD_ARGUMENTS);
+            if (key.empty() || key == "/")
+                throw Exception("Invalid S3 URI key: " + uri.toString(), ErrorCodes::BAD_ARGUMENTS);
```
Comment (alexey-milovidov) [inline dbms/src/IO/S3Common.cpp:135]: Need to provide explanation: "why exactly this S3 URI key is invalid, for what reason" in every exception message.

**Review by alexey-milovidov — APPROVED** [review]
Ok, only a few changes remain.

File: dbms/src/IO/S3Common.cpp (hunk 3 above)
Comment (Jokser) [reply dbms/src/IO/S3Common.cpp:108]: No, it means that there shouldbe "." or "-" after "s3"

Comment (alexey-milovidov) [reply dbms/src/IO/S3Common.cpp]: Ok, I've added another backslash for convenience.

**Comment by Jokser** [comment]
@alexey-milovidov Everything is ok. Could you please merge it?

**Pull request merged**